### PR TITLE
refactor!: eliminate remaining nested dicts usage

### LINKED ISSUE #282: Cleanup current solver code
Generally simplify the code and get rid of those deeply nested standard types (list, dict, tuple...)

- [ ] simplify csp solver code
- [ ] simplify validation code

### PR DESCRIPTION
**Pull request by spflueger**
Closes #282

### CODE REVIEW

**Review by redeboer — COMMENTED** [review]
I like it, good job!
Some comments and ideas, but no major changes

File: expertsystem/solving/__init__.py [hunk 1]
```diff
@@ -1122,24 +1006,23 @@ def __call__(
...
-            self.rule,
-            self.part_in,
-            self.part_out,
-            self.interaction_qns,
-            edge_qn_to_enum,
-        ):
-            self.conditions_never_met = True
-            return True
+        self.__update_variable_lists(params)
+        if str(self.rule) != "IsoSpinValidity":
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/solving/__init__.py:1010]: True

File: expertsystem/solving/__init__.py [hunk 2]
```diff
@@ -1150,20 +1033,27 @@ def __call__(
...
+                and qn_type in self.__out_edges_qns[index]
+            ):
+                self.__out_edges_qns[index][qn_type] = value  # type: ignore
+            elif qn_type in self.__node_qns:
+                self.__node_qns[qn_type] = value  # type: ignore
             else:
                 raise ValueError(
                     "The variable with name "
-                    + var_name
+                    + getattr(qn_type, "__name__")
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/solving/__init__.py:1057]: thats not defined for those NewTypes

File: expertsystem/state/properties.py [hunk 3]
```diff
@@ -16,37 +16,17 @@
     Union,
 )
 
-from numpy import arange
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/state/properties.py:19]: Yeah, can you open a new issue for this. Basically we just need one function which can create those spin domains

File: expertsystem/ui/_default_settings.py [hunk 4]
```diff
@@ -140,6 +163,7 @@ def create_default_interaction_settings(
             MuonLNConservation(),
             TauLNConservation(),
             BaryonNumberConservation(),
+            IsoSpinValidity(),  # should be changed to a pure edge rule
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/ui/_default_settings.py:166]: I just created some new issues.

File: expertsystem/solving/__init__.py [hunk 5]
```diff
@@ -668,6 +572,35 @@ def _create_variable_containers(
     return Result([graph])
 
 
+_EdgeVariableInfo = Tuple[int, Type[EdgeQuantumNumber]]
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/solving/__init__.py:575]: its the edge or node id, which is needed to map the csp variable to graph elements later on. The graph elements are fed to the rules

File: expertsystem/solving/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (redeboer) [reply expertsystem/solving/__init__.py:1057]: Sorry, I mean `qn_type.__name__`

File: expertsystem/ui/_system_control.py [hunk 6]
```diff
@@ -39,8 +37,8 @@
 
 def _change_qn_domain(
     settings: Tuple[EdgeSettings, NodeSettings],
-    qn_name: InteractionQuantumNumberNames,
-    new_domain: List[Spin],
+    qn_name: Type[NodeQuantumNumber],
+    new_domain: List[Union[int, float]],
```
[the remark this answers is not in the document]
Comment (redeboer) [reply expertsystem/ui/_system_control.py:41]: #305

**Comment by redeboer** [comment]
Oh forgot to mention: is `deepcopy` still the way to go? Particlularly with regard to the `Particle` isntances?

File: expertsystem/state/properties.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (redeboer) [reply expertsystem/state/properties.py:19]: #306

File: expertsystem/solving/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (redeboer) [reply expertsystem/solving/__init__.py:575]: Okay thanks

**Comment by spflueger** [comment]
> Oh forgot to mention: is `deepcopy` still the way to go? Particlularly with regard to the `Particle` isntances?

I tried to work with copy, but this does not find the correct solutions... I have to think about how we can speed this up, lets open an issue for this (question label first)

File: expertsystem/solving/__init__.py [hunk 7]
```diff
@@ -302,15 +276,13 @@ def _check_requirements(
         (in_edge_props, out_edge_props, node_props),
     ):
         if arg_counter == 3:
-            if not _check_arg_requirements(arg_type, props, key_mapping):  # type: ignore
+            if not _check_arg_requirements(arg_type, props):  # type: ignore
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/solving/__init__.py:279]: The props type is wrong according to mypy. Some how zip removes the type annotation. I don't know how I can get this to work

[the remark this answers is not in the document]
Comment (redeboer) [reply expertsystem/solving/__init__.py:279]: Ok interesting. Well, it's not that important. And those `# type: ignore` statements remain there as a reminder.

File: expertsystem/solving/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/solving/__init__.py]: 197acaa5488ebb20e6bd2c74e99ef76b8900a704
I actually dont understand how this access is allowed

File: expertsystem/ui/_system_control.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/ui/_system_control.py:41]: 197acaa5488ebb20e6bd2c74e99ef76b8900a704

File: expertsystem/ui/_default_settings.py [hunk 8]
```diff
@@ -85,6 +86,12 @@ def _get_ang_mom_magnitudes(is_nbody: bool) -> List[float]:
     return [0, 1, 2]
 
 
+def __create_projections(
+    magnitudes: List[Union[int, float]]
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/ui/_default_settings.py:90]: 197acaa5488ebb20e6bd2c74e99ef76b8900a704

File: expertsystem/solving/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (redeboer) [reply expertsystem/solving/__init__.py]: It's a [special member name](https://docs.python.org/3/reference/datamodel.html#specialnames), not dunder.

File: expertsystem/solving/__init__.py [hunk 9]
```diff
@@ -218,39 +205,31 @@ def _is_optional(class_field: Any) -> bool:
 
 
 def _init_class(
-    class_type: Type, props: Dict[Enum, Any], key_mapping: Optional[dict]
+    class_type: Type,
+    props: Union[
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/solving/__init__.py]: 197acaa5488ebb20e6bd2c74e99ef76b8900a704

File: expertsystem/solving/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/solving/__init__.py]: 197acaa5488ebb20e6bd2c74e99ef76b8900a704

**Review by redeboer — APPROVED** [review]
:+1:

**Pull request merged**